src/Skia/Avalonia.Skia/WriteableBitmapImpl.cs

```diff
@@ -31,44 +33,58 @@ public WriteableBitmapImpl(PixelSize size, Vector dpi, PixelFormat? format = nul
             Dpi = dpi;
 
             var colorType = PixelFormatHelper.ResolveColorType(format);
-            
+
             var runtimePlatform = AvaloniaLocator.Current?.GetService<IRuntimePlatform>();
-            
+
             if (runtimePlatform != null)
             {
+                var nfo = new SKImageInfo(size.Width, size.Height, colorType, SKAlphaType.Premul);
+
+                _blob = runtimePlatform.AllocBlob(nfo.BytesSize);
+
+                _pixMap = new SKPixmap(nfo, _blob.Address, nfo.RowBytes);
+
                 _bitmap = new SKBitmap();
 
-                var nfo = new SKImageInfo(size.Width, size.Height, colorType, SKAlphaType.Premul);
-                var blob = runtimePlatform.AllocBlob(nfo.BytesSize);
+                _bitmap.InstallPixels(_pixMap);
+
+                _bitmap.SetImmutable();
 
-                _bitmap.InstallPixels(nfo, blob.Address, nfo.RowBytes, s_releaseDelegate, blob);
+                _image = SKImage.FromBitmap(_bitmap);
             }
             else
             {
                 _bitmap = new SKBitmap(size.Width, size.Height, colorType, SKAlphaType.Premul);
+
+                _pixMap = _bitmap.PeekPixels();
+
+                _image = SKImage.FromBitmap(_bitmap);
             }
 
-            _bitmap.Erase(SKColor.Empty);
+            _pixMap.Erase(SKColor.Empty);
         }
 
         public Vector Dpi { get; }
 
         /// <inheritdoc />
         public PixelSize PixelSize { get; }
 
-        public int Version { get; private set; } = 1;
+        public int Version { get; internal set; } = 1;
 
         /// <inheritdoc />
         public void Draw(DrawingContextImpl context, SKRect sourceRect, SKRect destRect, SKPaint paint)
         {
             lock (_lock)
-                context.Canvas.DrawBitmap(_bitmap, sourceRect, destRect, paint);
+                context.Canvas.DrawImage(_image, sourceRect, destRect, paint);
         }
 
         /// <inheritdoc />
         public void Dispose()
         {
+            _image.Dispose();
```

Comment:
Keep in mind that this code might be never called for garbage collected instances. That was one of the reasons for a custom memory release callback

Reply:
The callback was only executed when the bitmap was disposed how is that any different?

Reply:
It is also executed when bitmap is *not* disposed. In that case SKBitmap finalizer destroys the native SKBitmap instance which in turn calls the callback.